drivers/rtc/rtc_rv8803.c

```diff
@@ -0,0 +1,868 @@
+	}
+
+	/* Update WADA bit */
+	if ((mask & RTC_ALARM_TIME_MASK_MONTHDAY) || (mask & RTC_ALARM_TIME_MASK_WEEKDAY)) {
+		reg_val = (mask & RTC_ALARM_TIME_MASK_MONTHDAY) ? RV8803_EXTENSION_WADA_BIT : 0;
+		err = rv8803_update_reg8(dev, RV8803_EXTENSION_REG, RV8803_EXTENSION_WADA_BIT,
+					 reg_val);
+		if (err) {
+			return err;
+		}
+	}
+
+	/* Update alarm registers */
+	err = rv8803_write_regs(dev, RV8803_MINUTES_ALARM_REG, regs, sizeof(regs));
+	if (err) {
+		return err;
+	}
+
+	LOG_DBG("Set alarm: month day: %d, week day: %d, hour: %d, minute: %d, mask: 0x%04X",
+		timeptr->tm_mday, timeptr->tm_wday, timeptr->tm_hour, timeptr->tm_min, mask);
+
+	return 0;
+}
+
+static int rv8803_alarm_get_time(const struct device *dev, uint16_t id, uint16_t *mask,
+				 struct rtc_time *timeptr)
+{
+	uint8_t regs[3];
+	uint8_t reg_val;
+	int err;
+
+	if (id != 0) {
+		LOG_ERR("Invalid alarm ID: %d", id);
+		return -EINVAL;
+	}
+
+	err = rv8803_read_regs(dev, RV8803_MINUTES_ALARM_REG, regs, sizeof(regs));
+	if (err) {
+		return err;
+	}
+
+	/* Read extension register to get WADA bit */
+	err = rv8803_read_reg8(dev, RV8803_EXTENSION_REG, &reg_val);
+	if (err) {
+		return err;
+	}
+
+	memset(timeptr, 0, sizeof(*timeptr));
+	*mask = 0;
+
+	if ((regs[0] & RV8803_MINUTES_ALARM_AE_M_BIT) == 0) {
+		timeptr->tm_min = bcd2bin(regs[0] & RV8803_MINUTES_ALARM_MASK);
+		*mask |= RTC_ALARM_TIME_MASK_MINUTE;
+	}
+
+	if ((regs[1] & RV8803_HOURS_ALARM_AE_H_BIT) == 0) {
+		timeptr->tm_hour = bcd2bin(regs[1] & RV8803_HOURS_ALARM_MASK);
+		*mask |= RTC_ALARM_TIME_MASK_HOUR;
+	}
+
+	if ((regs[2] & RV8803_WEEKDAY_OR_DATE_ALARM_AE_WD_BIT) == 0) {
+		if (reg_val & RV8803_EXTENSION_WADA_BIT) {
+			timeptr->tm_mday = bcd2bin(regs[2] & RV8803_DATE_ALARM_MASK);
+			*mask |= RTC_ALARM_TIME_MASK_MONTHDAY;
+		} else {
+			timeptr->tm_wday = find_lsb_set(regs[2] & RV8803_WEEKDAY_ALARM_MASK);
+			*mask |= RTC_ALARM_TIME_MASK_WEEKDAY;
+		}
+	}
+
+	LOG_DBG("Get alarm: month day: %d, week day: %d, hour: %d, minute: %d, mask: 0x%04X",
+		timeptr->tm_mday, timeptr->tm_wday, timeptr->tm_hour, timeptr->tm_min, *mask);
+
+	return 0;
+}
+
+static int rv8803_alarm_is_pending(const struct device *dev, uint16_t id)
+{
+	uint8_t flags;
+	int err;
+
+	if (id != 0) {
+		LOG_ERR("Invalid alarm ID: %d", id);
+		return -EINVAL;
+	}
+
+	rv8803_lock_sem(dev);
+
+	do {
+		err = rv8803_read_reg8(dev, RV8803_FLAG_REG, &flags);
+		if (err) {
+			break;
+		}
+
+		if (flags & RV8803_FLAG_AF_BIT) {
+			flags &= ~RV8803_FLAG_AF_BIT;
+
+			err = rv8803_write_reg8(dev, RV8803_FLAG_REG, flags);
+			if (err) {
+				break;
+			}
+
+			/* Indicate that alarm is pending */
+			err = 1;
+		}
+	} while (0);
+
+	rv8803_unlock_sem(dev);
+
+	return err;
+}
+
+static int rv8803_alarm_set_callback(const struct device *dev, uint16_t id,
+				     rtc_alarm_callback callback, void *user_data)
+{
+#ifndef RV8803_INT_GPIOS_IN_USE
+	ARG_UNUSED(dev);
+	ARG_UNUSED(id);
+	ARG_UNUSED(callback);
+	ARG_UNUSED(user_data);
+
+	return -ENOTSUP;
```

Comment:
not implemented is actually `-ENOSYS`, btw `z_impl_rtc_alarm_set_callback` checks for that so another option would be to not define the function at all

Reply:
I've taken this part from another MicroCrystal RTC driver: https://github.com/zephyrproject-rtos/zephyr/blob/e573b6bef964e06680c82ee7aeae2b2f6f49ef29/drivers/rtc/rtc_rv3028.c#L691 here `-ENOTSUP` value is used. Should I use `-ENOSYS` instead?

Reply:
yeah, it doesn't really make much of a difference in practice but that's what's normally used in the API wrappers, see `include/zephyr/drivers/rtc.h`. IMO though it'd be bettere to not build it at all, save few bytes of flash from the wrapper.

Reply:
Thanks for the explanation, fixed as per your suggestions